qubes/storage/reflink.py

```diff
@@ -0,0 +1,434 @@
+#
+
+''' Driver for handling VM images as files, without any device-mapper
+    involvement. A reflink-capable filesystem is strongly recommended,
+    but not required.
+'''
+
+import collections
+import errno
+import fcntl
+import glob
+import logging
+import os
+import re
+import subprocess
+import tempfile
+from contextlib import contextmanager, suppress
+
+import qubes.storage
+
+BLKSIZE = 512
+FICLONE = 1074041865  # see ioctl_ficlone manpage
+LOGGER = logging.getLogger('qube.storage.reflink')
+
+
+class ReflinkPool(qubes.storage.Pool):
+    driver = 'file-reflink'
+    _known_dir_path_prefixes = ['appvms', 'vm-templates']
+
+    def __init__(self, dir_path, setup_check='yes', revisions_to_keep=1,
+                 **kwargs):
+        super(ReflinkPool, self).__init__(revisions_to_keep=revisions_to_keep,
+                                          **kwargs)
+        self._volumes = {}
+        self.dir_path = os.path.abspath(dir_path)
+        self.setup_check = qubes.property.bool(None, None, setup_check)
+
+    def setup(self):
+        created = _make_dir(self.dir_path)
+        if self.setup_check and not is_reflink_supported(self.dir_path):
+            if created:
+                _remove_empty_dir(self.dir_path)
+            raise qubes.storage.StoragePoolException(
+                'The filesystem for {!r} does not support reflinks. If you'
+                ' can live with VM startup delays and wasted disk space, pass'
+                ' the "setup_check=no" option.'.format(self.dir_path))
+        for dir_path_prefix in self._known_dir_path_prefixes:
+            _make_dir(os.path.join(self.dir_path, dir_path_prefix))
+        return self
+
+    def init_volume(self, vm, volume_config):
+        # Fail closed on any strange VM dir_path_prefix, just in case
+        # /etc/udev/rules/00-qubes-ignore-devices.rules needs updating
+        assert vm.dir_path_prefix in self._known_dir_path_prefixes, \
+               'Unknown dir_path_prefix {!r}'.format(vm.dir_path_prefix)
+
+        volume_config['pool'] = self
+        if 'revisions_to_keep' not in volume_config:
+            volume_config['revisions_to_keep'] = self.revisions_to_keep
+        if 'vid' not in volume_config:
+            volume_config['vid'] = os.path.join(vm.dir_path_prefix, vm.name,
+                                                volume_config['name'])
+        if volume_config['name'] == 'root':
+            # When root is read-only, the VM itself will set up a
+            # device-mapper CoW layer at /dev/mapper/dmroot, writing
+            # modified blocks to the volatile volume. That volume
+            # usually does CoW on our layer too, thus creating an
+            # inefficient CoW-on-CoW setup. Avoid this by always
+            # overriding root to be read-write - which may become
+            # incompatible with a future untrusted storage domain!
+            volume_config['rw'] = True
```

Comment:
Better patch it directly where [volume_config is defined](https://github.com/qubesos/qubes-core-admin/blob/master/qubes/vm/appvm.py#L48) (and similar for other VM types).

Reply:
And please separate commit for such changes.

Reply:
[Done.](https://github.com/QubesOS/qubes-core-admin/pull/188/commits/958d6c7067138f82e602b18ee8849bbe7ef0a565)